torch/csrc/autograd/engine.cpp

```diff
@@ -113,134 +123,129 @@ auto Engine::thread_on_exception(FunctionTask& task, std::exception& e) -> void
     std::stringstream ss;
-    ss << "Function '" << fn.name() << "' returned an invalid number of gradients - expected ";
-    ss << fn.previous_functions.size() << ", but got " << grad_inputs.size();
+    ss << "Function '" << fn.name() << "' returned an invalid number of outputs - expected ";
+    ss << fn.next_functions.size() << ", but got " << outputs.size();
     throw std::runtime_error(ss.str());
   }
 
-  int size = grad_inputs.size();
-  for (int i = 0; i < size; ++i) {
-    auto& grad_input = grad_inputs[i];
-    auto& prev_fn = fn.previous_functions[i].first;
-    int output_nr = fn.previous_functions[i].second;
+  int num_outputs = outputs.size();
+  for (int i = 0; i < num_outputs; ++i) {
+    auto& output = outputs[i];
+    auto& next_fn = fn.next_functions[i].first;
+    int input_nr = fn.next_functions[i].second;
 
-    // null inputs have no previous_function and we skip them here
-    if (!prev_fn) {
+    if (!next_fn) {
       continue;
     }
 
     // Stochastic functions are placed in the ready queue by
-    // compute_dependencies, so we can skip them here.
-    if (prev_fn->is_stochastic || !prev_fn->requires_grad) {
+    // compute_dependencies, so we have to skip them here.
+    if (next_fn->is_stochastic || !next_fn->is_executable) {
       continue;
     }
 
     std::lock_guard<std::mutex> lock(task.base->mutex);
-    if (auto var = dynamic_cast<Variable*>(prev_fn.get())) {
-      if (!grad_input) {
-        // NOTE: grad_input can be NULL if the function returns None for a
-        // non_differentiable input. We may need to track additional information
-        // at the function level to determine if a NULL grad_input is an error.
-        std::stringstream ss;
-        ss << "Function '" << fn.name() << "' missing gradient at " << i;
-        throw std::runtime_error(ss.str());
-      }
-      var->backward(grad_input);
-      continue;
-    }
-
-    // Check if the function is ready for backward
+    // Check if the next function is ready to be computed
     bool is_ready = false;
     auto& dependencies = task.base->dependencies;
-    auto it = dependencies.find(prev_fn.get());
+    auto it = dependencies.find(next_fn.get());
     if (it == dependencies.end()) {
-      auto name = prev_fn->name();
+      auto name = next_fn->name();
       throw std::runtime_error(std::string("dependency not found for ") + name);
     } else if (--it->second == 0) {
       dependencies.erase(it);
       is_ready = true;
     }
 
     auto& not_ready = task.base->not_ready;
-    auto not_ready_it = not_ready.find(prev_fn.get());
+    auto not_ready_it = not_ready.find(next_fn.get());
     if (not_ready_it == not_ready.end()) {
       // No buffers have been allocated for the function
-      GradBuffer prev_buffer(prev_fn->num_outputs);
-      prev_buffer.addGrad(output_nr, std::move(grad_input));
+      InputBuffer input_buffer(next_fn->num_inputs);
+      input_buffer.add(input_nr, std::move(output));
```

Comment:
This should be protected by an `input_nr >= 0` or `next_fn->num_inputs > 0` because in the case for example of the `Error()` function, we try to write the output in an InputBuffer of size 0 (which segfauts).